### PR TITLE
Fix #847: enable back GD_BICUBIC* interpolation methods

### LINKED ISSUE #847: gdImageSetInterpolationMethod ignores GD_BICUBIC*
**Describe the bug**
`gdImageSetInterpolationMethod` does not longer accept `GD_BICUBIC` or `GD_BICUBIC_FIXED` constants.

As such, the following code snippet do not scale image with bicubic algorithms:

```
if (gdImageSetInterpolationMethod(im, algo)) {
  scaled = gdImageScale(im, width2, height2);
}
```

The change was introduced in 672475e540da84100252e720291c63b30e028d06.

**To Reproduce**
```c
#include "gd.h"
//#include "gdtest.h

int main()
{
  gdImagePtr im;
  int status;
  im = gdImageCreate(1, 1);
  gdImageSetInterpolationMethod(im, GD_BICUBIC_FIXED);
  if (gdImageGetInterpolationMethod(im) == GD_BICUBIC_FIXED) {
    status = 0;
  } else {
    status = 1;
  }
  // if (status > 0) {
  //   gdTestErrorMsg("expecting success return value\n");
  // }
  gdImageDestroy(im);
  return status;
}
```

**Expected behavior**
`gdImageSetInterpolationMethod` should set the mode and return 1 **OR** the ChangeLog should mention that `GD_BICUBIC*` are removed and ignored.

**Actual results**
`gdImageSetInterpolationMethod` returns 0, and code guarding a call to `gdImageScale` with `if (gdImageSetInterpolationMethod(im, GD_BICUBIC)) { … }` is silently ignored, this is what PHP-gd does: https://github.com/php/php-src/blob/master/ext/gd/gd.c#L3643

**Environment (please complete the following information):**
 - OS: Debian testing (with debian's libgd) or any Debian using the [deb.sury.org](https://github.com/oerdnj/deb.sury.org) repository with php-gd.
 - Version 2.3.3 
- All libraries used by GD and their respective version : not relevant here

**Additional context**
`IMG_BICUBIC` and `IMG_BICUBIC_FIXED` are mentioned and promoted on php-gd documentation for `imagescale` : https://www.php.net/manual/en/function.imagescale.php

We have some legacy code that use `IMG_BICUBIC_FIXED` and suddenly stop working after upgrading from debian 9 to debian 10 with php-sury repository which ships libgd3 2.3.3 instead of 2.2.5.

### PR DESCRIPTION
**Pull request by fulax**
I made the PR for the GD-2.3 branch, as 2.3.3 is the version with the regression.

### CODE REVIEW

File: tests/gdimagesetinterpolationmethod/github_bug_00847.c [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  im = gdImageCreate(1, 1);
+  gdImageSetInterpolationMethod(im, GD_BICUBIC_FIXED);
+  if (gdImageGetInterpolationMethod(im) == GD_BICUBIC_FIXED) {
+    status = 0;
+  } else {
+    status = 1;
+  }
+  if (status > 0) {
+    gdTestErrorMsg("expecting success return value\n");
+  }
```
Comment (cmb69) [inline tests/gdimagesetinterpolationmethod/github_bug_00847.c:17]: This looks unnecessarily complicated. Instead we could just do:
```suggestion
  gdTestAssert(gdImageGetInterpolationMethod(im) == GD_BICUBIC_FIXED);
```

File: tests/gdimagesetinterpolationmethod/github_bug_00847.c [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  gdImageDestroy(im);
+  return status;
```
Comment (cmb69) [inline tests/gdimagesetinterpolationmethod/github_bug_00847.c:19]: For the above modifications to have effect, we should do:
```suggestion
  return gdNumFailures();
```

File: tests/gdimagesetinterpolationmethod/github_bug_00847.c [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#include "gd.h"
+#include "gdtest.h"
+
+int main()
+{
+  gdImagePtr im;
+  int status;
```
Comment (cmb69) [inline tests/gdimagesetinterpolationmethod/github_bug_00847.c:7]: This line is no longer necessary, if the other changes are done:
```suggestion
```

**Review by cmb69 — COMMENTED** [review]
Thank you for the PR! The actual fix looks fine to me; like I said, I have some suggestions for the accompanying test case.

**Comment by cmb69** [comment]
Hmm, the macOS failure is unrelated to this PR, but should be catered to nonetheless (not necessarily by you, @fulax). Apparently, we have the libpng version hard-coded to 1.6.37, but now we need to use 1.6.38. Would anybody know whether we can get rid of the actual version number? Otherwise, I'd make sense to use a variable for the version (so it needs to be changed only in one place).

**Comment by fulax** [comment]
> Thank you for the PR! The actual fix looks fine to me; like I said, I have some suggestions for the accompanying test case.

All applied, thanks. I was not familiar with the test API, I just looked at some random test file and build from that one.

**Comment by fulax** [comment]
> Would anybody know whether we can get rid of the actual version number?

I'm also not familiar with MacOS build environments, but the actual config seems to rely on pkg-config, so I guess `pkg-config --modversion libpng16` could be use to get that version number.

**Comment by cmb69** [comment]
> I was not familiar with the test API, I just looked at some random test file and build from that one.

Yeah, we should have some documentation about the test suite. Until then, the best available info is in [gdtest.h](https://github.com/libgd/libgd/blob/master/tests/gdtest/gdtest.h).

> I'm also not familiar with MacOS build environments, but the actual config seems to rely on pkg-config, so I guess `pkg-config --modversion libpng16` could be use to get that version number.

Ah, that seems to be a good idea.

Anyhow, you also need to add the new test case to the respective [CMakeLists.txt](https://github.com/libgd/libgd/blob/master/tests/gdimagesetinterpolationmethod/CMakeLists.txt); currently, the test is apparently not run on CI.

**Review by cmb69 — APPROVED** [review]
Thank you! This looks good to me.

@vapier, what is our current merge strategy – target GD-2.3 or master?

**Comment by fulax** [comment]
Hello,

Any chance we can go on with this issue? I could rewrite the patch or change the target branch if needed.

Let me know,
Thanks.

**Comment by vapier** [comment]
is the issue fixed on master already ?  if not, we always merge to master first, and then cherry pick back to older branches if we want to fix it there.

**Comment by cmb69** [comment]
Thanks @vapier (wasn't sure anymore).

@fulax, I've changed the branch, but that shows unrelated commits, so please rebase the PR onto current "master", force-push, and change the branch in the GH UI.

**Comment by cmb69** [comment]
The test failures are unrelated; I'm going to merge. Thank you for the PR!

**Comment by vapier** [comment]
at this point, we usually just cherry pick the fix back ourselves, although we don't say no to PRs in general.  which i've done now.

**Pull request merged**